File: android/app/src/main/java/ca/psiphon/conduit/nativemodule/ConduitStateService.java
package ca.psiphon.conduit.nativemodule;

import android.app.Service;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Binder;
import android.os.DeadObjectException;
import android.os.IBinder;
import android.os.RemoteException;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import ca.psiphon.conduit.nativemodule.logging.MyLog;
import ca.psiphon.conduit.state.IConduitStateCallback;
import ca.psiphon.conduit.state.IConduitStateService;
import io.reactivex.Flowable;
import io.reactivex.disposables.CompositeDisposable;

public class ConduitStateService extends Service {

    private static final String TAG = ConduitStateService.class.getSimpleName();

    private record StateUpdate(int appVersion, ProxyState proxyState) {
        // Current schema version for the state update JSON structure
        private static final int CURRENT_SCHEMA = 0;

        String toJson() {
            // Wraps state data with schema version to support future changes
            // to the JSON structure without breaking clients.
            // JSON structure:
            // {
            //  "schema": 1, // Current schema version
            //  "data": {
            //    "appVersion": 123, // App version code
            //    "running": true/false, // Proxy running status, omitted for UNKNOWN state
            //  }
            JSONObject data = new JSONObject();
            JSONObject wrapper = new JSONObject();
            try {
                data.put("appVersion", appVersion);
                // For UNKNOWN proxy state, omit running status since it represents an uninitialized state
                if (!proxyState.isUnknown()) {
                    data.put("running", proxyState.isRunning());
                }
                // Only schema version at top level,
                wrapper.put("schema", CURRENT_SCHEMA);
                wrapper.put("data", data);
            } catch (JSONException e) {
                MyLog.e(TAG, "Failed to create JSON object: " + e.getMessage());
            }
            return wrapper.toString();
        }
    }

    private final CompositeDisposable compositeDisposable = new CompositeDisposable();

    // Map to hold registered clients and their subscriptions
    private final Map<IBinder, IConduitStateCallback> clients = new ConcurrentHashMap<>();
    // Lock for clients map access
    private final Object clientsLock = new Object();

    // Interactor for getting state from the ConduitService
    private ConduitServiceInteractor conduitServiceInteractor;

    // Flowable for the running state of the service in JSON format
    // This Flowable is updated whenever the service state changes
    // and is used to update all registered clients
    private Flowable<String> runningState;

    // Holds current state update to send to newly registered clients
    private StateUpdate currentUpdate = null;

    // AIDL binder implementation
    private final IConduitStateService.Stub binder = new IConduitStateService.Stub() {
        @Override
        public void registerClient(IConduitStateCallback client) {
            if (client == null) {
                return;
            }

            // Check if the client is trusted (do this outside the lock since it doesn't need synchronization)
            int uid = Binder.getCallingUid();
            if (!isTrustedUid(uid)) {
                throw new SecurityException("Client is not authorized to register with this service.");
            }

            synchronized (clientsLock) {
                IBinder clientBinder = client.asBinder();
                if (!clients.containsKey(clientBinder)) {
                    clients.put(clientBinder, client);
                    if (currentUpdate != null) {
                        try {
                            client.onStateUpdate(currentUpdate.toJson());
                        } catch (RemoteException e) {
                            MyLog.e(TAG, "Failed to notify client: " + clientBinder + ", " + e.getMessage());
                        }
                    }
                    MyLog.i(TAG, "Client registered: " + clientBinder);
                }
            }
        }

        @Override
        public void unregisterClient(IConduitStateCallback client) {
            if (client == null) {
                return;
            }

            synchronized (clientsLock) {
                IBinder clientBinder = client.asBinder();
                clients.remove(clientBinder);
                MyLog.i(TAG, "Client unregistered: " + clientBinder);
            }
        }
    };

    @Override
    public void onCreate() {
        MyLog.init(getApplicationContext());

        // Load runtime trusted signatures configuration from file
        PackageHelper.configureRuntimeTrustedSignatures(PackageHelper.readTrustedSignaturesFromFile(getApplicationContext()));

        conduitServiceInteractor = new ConduitServiceInteractor(getApplicationContext());
        conduitServiceInteractor.onStart(getApplicationContext());

        initializeRunningState();
    }

    private void initializeRunningState() {
        runningState = Flowable.combineLatest(
                        Flowable.just(getAppVersionCode()),
                        conduitServiceInteractor.proxyStateFlowable().startWith(ProxyState.unknown()),
                        StateUpdate::new
                )
                .map(stateUpdate -> {
                    // Record the current state update to send to newly registered clients
                    currentUpdate = stateUpdate;
                    return stateUpdate.toJson();
                })
                .distinctUntilChanged();

        // Single subscription to the runningState Flowable to update all registered clients
        compositeDisposable.add(runningState.subscribe(
                state -> {
                    synchronized (clientsLock) {
                        for (Map.Entry<IBinder, IConduitStateCallback> entry : clients.entrySet()) {
                            IBinder clientBinder = entry.getKey();
                            IConduitStateCallback client = entry.getValue();
                            try {
                                client.onStateUpdate(state);
                            } catch (RemoteException e) {
                                // Remove the client if it is dead and do not log the exception as it is expected
                                // to happen when a client goes away without unregistering.
                                if (e instanceof DeadObjectException) {
                                    clients.remove(clientBinder);
                                } else {
                                    MyLog.e(TAG, "Failed to notify client: " + clientBinder + ", " + e.getMessage());
                                }
                            }
                        }
                    }
                },
                throwable -> MyLog.e(TAG, "Error in runningState flow: " + throwable.getMessage())
        ));
    }

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public void onDestroy() {
        compositeDisposable.dispose();
        synchronized (clientsLock) {
            clients.clear();
        }
        conduitServiceInteractor.onStop(getApplicationContext());
        conduitServiceInteractor.onDestroy(getApplicationContext());
    }

    private int getAppVersionCode() {
        try {
            return getPackageManager()
                    .getPackageInfo(getPackageName(), 0)
                    .versionCode;
        } catch (PackageManager.NameNotFoundException e) {
            MyLog.e(TAG, "Failed to fetch app version code: " + e.getMessage());
            return -1;
        }
    }

    // Check if the calling UID is trusted
    private boolean isTrustedUid(int uid) {
        // Get the package names associated with the calling UID
        String[] packages = getPackageManager().getPackagesForUid(uid);

        if (packages == null || packages.length == 0) {
            MyLog.e(TAG, "Calling UID has no associated packages, rejecting.");
            return false;
        }

        // It is possible to have multiple packages associated with the same UID, iterate through all
        for (String packageName : packages) {
            if (PackageHelper.verifyTrustedPackage(getPackageManager(), packageName)) {
                return true;
            }
        }
        // Reject the UID if none of the packages are trusted
        MyLog.w(TAG, "None of the associated packages were trusted, rejecting UID.");
        return false;
    }
}

File: modules/expo-psiphon-tunnel-core/android/src/main/java/expo/modules/psiphontunnelcore/ConduitStateService.kt
/*
 * Copyright (c) 2026, Psiphon Inc.
 * All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package expo.modules.psiphontunnelcore

import android.app.Service
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Binder
import android.os.Bundle
import android.os.DeadObjectException
import android.os.IBinder
import android.os.RemoteException
import ca.psiphon.conduit.nativemodule.IConduitClientCallback
import ca.psiphon.conduit.nativemodule.IConduitService
import ca.psiphon.conduit.state.IConduitStateCallback
import ca.psiphon.conduit.state.IConduitStateService
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap

class ConduitStateService : Service() {
    companion object {
        private const val TAG = "ConduitStateService"
        private const val SCHEMA_VERSION = 1
        private const val BIND_ACTION = "ca.psiphon.conduit.ACTION_BIND_CONDUIT_STATE"
    }

    private enum class ProxyState {
        RUNNING,
        STOPPED,
        UNKNOWN,
    }

    private data class StateUpdate(
        val appVersion: Int,
        val state: ProxyState,
    ) {
        fun toJson(): String {
            val data = JSONObject().put("appVersion", appVersion)
            if (state != ProxyState.UNKNOWN) {
                data.put("running", state == ProxyState.RUNNING)
            }
            return JSONObject()
                .put("schema", SCHEMA_VERSION)
                .put("data", data)
                .toString()
        }
    }

    private val clients = ConcurrentHashMap<IBinder, IConduitStateCallback>()
    private val clientsLock = Any()
    private var inproxyService: IConduitService? = null
    private var isInproxyServiceBound = false
    private var isDestroyed = false
    @Volatile
    private var currentUpdate = StateUpdate(
        appVersion = -1,
        state = ProxyState.UNKNOWN,
    )

    private fun logInfo(message: String) {
        AppLogStore.info(applicationContext, TAG, message)
    }

    private fun logWarn(message: String) {
        AppLogStore.warn(applicationContext, TAG, message)
    }

    private fun logError(message: String, error: Throwable) {
        AppLogStore.error(applicationContext, TAG, message, error)
    }

    private val inproxyClientCallback = object : IConduitClientCallback.Stub() {
        override fun onProxyStateUpdated(proxyStateBundle: Bundle) {
            updateAndNotify(proxyStateFromBundle(proxyStateBundle))
        }

        override fun onProxyActivityStatsUpdated(proxyActivityStatsBundle: Bundle) {
            // The external state API only mirrors whether the proxy is running.
        }

        override fun onProxyError(proxyErrorBundle: Bundle) {
            // Errors are delivered to the app UI through ExpoPsiphonTunnelCoreModule.
        }
    }

    private val inproxyServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            logInfo("Connected to InproxyForegroundService")
            inproxyService = IConduitService.Stub.asInterface(service)
            try {
                inproxyService?.registerClient(inproxyClientCallback)
            } catch (error: RemoteException) {
                logError("Failed to register inproxy state callback", error)
            }
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            logInfo("Disconnected from InproxyForegroundService")
            inproxyService = null
            isInproxyServiceBound = false
            if (!isDestroyed) {
                bindInproxyService()
            }
        }
    }

    private val binder = object : IConduitStateService.Stub() {
        override fun registerClient(client: IConduitStateCallback?) {
            if (client == null) {
                return
            }

            val caller = enforceTrustedCaller("registerClient")
            logInfo("Accepted registerClient from $caller")

            val clientBinder = client.asBinder()
            var activeClientCount = 0
            synchronized(clientsLock) {
                clients[clientBinder] = client
                activeClientCount = clients.size
            }
            val payload = currentUpdate.toJson()
            try {
                client.onStateUpdate(payload)
            } catch (e: RemoteException) {
                if (e is DeadObjectException) {
                    synchronized(clientsLock) {
                        clients.remove(clientBinder, client)
                        activeClientCount = clients.size
                    }
                } else {
                    logError("Failed to deliver initial state", e)
                }
            }
            emitIpcEvent(
                type = "registerClient",
                status = "accepted",
                caller = caller,
                activeClientCount = activeClientCount,
            )
        }

        override fun unregisterClient(client: IConduitStateCallback?) {
            if (client == null) {
                return
            }
            var activeClientCount = 0
            synchronized(clientsLock) {
                clients.remove(client.asBinder())
                activeClientCount = clients.size
            }
            logInfo("Accepted unregisterClient")
            emitIpcEvent(
                type = "unregisterClient",
                status = "accepted",
                activeClientCount = activeClientCount,
            )
        }

        override fun fetchConduitPrivateKey(): String {
            val caller = enforceTrustedCaller("fetchConduitPrivateKey")
            logInfo("Accepted fetchConduitPrivateKey from $caller")

            val privateKey = InproxyParameters.load(applicationContext)?.privateKey.orEmpty()
            if (privateKey.isBlank()) {
                emitIpcEvent(
                    type = "fetchConduitPrivateKey",
                    status = "failed",
                    caller = caller,
                    message = "Conduit private key is not set",
                )
                throw IllegalStateException("Conduit private key is not set")
            }
            emitIpcEvent(
                type = "fetchConduitPrivateKey",
                status = "accepted",
                caller = caller,
            )
            return privateKey
        }
    }

    override fun onCreate() {
        super.onCreate()
        isDestroyed = false

        val fileTrustedSignatures =
            PackageHelper.readTrustedSignaturesFromFile(applicationContext)
        val devTrustedSignatures =
            PackageHelper.parseTrustedSignaturesJson(
                BuildConfig.DEV_TRUSTED_SIGNATURES_JSON,
            )

        PackageHelper.configureRuntimeTrustedSignatures(
            PackageHelper.mergeTrustedSignatures(
                fileTrustedSignatures,
                devTrustedSignatures,
            ),
        )

        if (devTrustedSignatures.isNotEmpty()) {
            logWarn("Loaded development IPC signatures for ${devTrustedSignatures.size} package(s).")
        }

        currentUpdate = StateUpdate(
            appVersion = appVersionCode(),
            state = if (Utils.getServiceRunningFlag(applicationContext)) {
                ProxyState.RUNNING
            } else {
                ProxyState.STOPPED
            },
        )

        bindInproxyService()
    }

    override fun onDestroy() {
        super.onDestroy()
        isDestroyed = true
        unbindInproxyService()
        synchronized(clientsLock) {
            clients.clear()
        }
    }

    override fun onBind(intent: Intent?): IBinder? {
        if (intent?.action != BIND_ACTION) {
            logWarn("Denying bind with invalid action: ${intent?.action}")
            emitIpcEvent(
                type = "bind",
                status = "invalid",
                message = "Invalid bind action: ${intent?.action}",
            )
            return null
        }
        // The framework invokes onBind during service setup, so Binder.getCallingUid()
        // here does not reliably identify the eventual external client. Enforce caller
        // authorization on the AIDL methods instead, where the remote UID is correct.
        logInfo("Accepted bind for action $BIND_ACTION")
        emitIpcEvent(
            type = "bind",
            status = "accepted",
            message = "Bind action accepted; caller authorization is enforced per IPC method",
        )
        return binder
    }

    private fun updateAndNotify(state: ProxyState) {
        val update = StateUpdate(
            appVersion = appVersionCode(),
            state = state,
        )
        if (update == currentUpdate) {
            return
        }
        currentUpdate = update

        val payload = update.toJson()
        val callbacks = synchronized(clientsLock) {
            clients.entries.map { it.key to it.value }
        }
        val toRemove = mutableListOf<IBinder>()
        callbacks.forEach { (clientBinder, callback) ->
            try {
                callback.onStateUpdate(payload)
            } catch (error: RemoteException) {
                if (error is DeadObjectException) {
                    toRemove.add(clientBinder)
                } else {
                    logError("Failed to notify state client", error)
                }
            }
        }
        if (toRemove.isEmpty()) {
            return
        }
        val activeClientCount = synchronized(clientsLock) {
            callbacks.forEach { (clientBinder, callback) ->
                if (clientBinder in toRemove) {
                    clients.remove(clientBinder, callback)
                }
            }
            clients.size
        }
        emitIpcEvent(
            type = "stateClient",
            status = "disconnected",
            activeClientCount = activeClientCount,
            message = "Removed disconnected state client callback",
        )
    }

    private fun proxyStateFromBundle(eventData: Bundle): ProxyState {
        return when (eventData.getString("status")) {
            "RUNNING" -> ProxyState.RUNNING
            "STOPPED" -> ProxyState.STOPPED
            else -> ProxyState.UNKNOWN
        }
    }

    private fun bindInproxyService() {
        if (isInproxyServiceBound) {
            return
        }
        val intent = Intent(applicationContext, InproxyForegroundService::class.java)
        isInproxyServiceBound = bindService(intent, inproxyServiceConnection, Context.BIND_AUTO_CREATE)
        if (!isInproxyServiceBound) {
            logWarn("bindService returned false for InproxyForegroundService")
        }
    }

    private fun unbindInproxyService() {
        try {
            inproxyService?.unregisterClient(inproxyClientCallback)
        } catch (error: RemoteException) {
            logError("Failed to unregister inproxy state callback", error)
        }
        try {
            if (isInproxyServiceBound) {
                unbindService(inproxyServiceConnection)
            }
        } catch (_: IllegalArgumentException) {
        }
        inproxyService = null
        isInproxyServiceBound = false
    }

    private fun isTrustedUid(uid: Int): Boolean {
        return PackageHelper.verifyTrustedCallingUid(applicationContext, uid)
    }

    private fun enforceTrustedCaller(operation: String): String {
        val uid = Binder.getCallingUid()
        val caller = PackageHelper.describeCallingUid(applicationContext, uid)
        if (!isTrustedUid(uid)) {
            logDeniedCaller(operation, caller)
            throw SecurityException("Client is not authorized: $operation")
        }
        return caller
    }

    private fun logDeniedCaller(operation: String, caller: String) {
        logWarn("Denied $operation from $caller")
        emitIpcEvent(
            type = operation,
            status = "denied",
            caller = caller,
            message = "Client is not authorized",
        )
    }

    private fun emitIpcEvent(
        type: String,
        status: String,
        caller: String? = null,
        activeClientCount: Int? = null,
        message: String? = null,
    ) {
        val data = Bundle().apply {
            putString("status", status)
            putLong("timestampMs", System.currentTimeMillis())
            if (!caller.isNullOrBlank()) {
                putString("caller", caller)
            }
            if (activeClientCount != null) {
                putInt("activeClientCount", activeClientCount)
            }
            if (!message.isNullOrBlank()) {
                putString("message", message)
            }
        }
        IpcEventQueue.enqueue(type, data)
    }

    private fun appVersionCode(): Int {
        return try {
            val packageInfo = packageManager.getPackageInfo(packageName, 0)
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
                packageInfo.longVersionCode.toInt()
            } else {
                @Suppress("DEPRECATION")
                packageInfo.versionCode
            }
        } catch (e: Exception) {
            logError("Failed to fetch app version code", e)
            -1
        }
    }
}
